src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -0,0 +1,183 @@
+        this.threadPool = threadPool;
+        timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, newValue -> timeout = newValue);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> indexingBulkSize = newValue);
+    }
+
+    @Override
+    protected void doExecute(final Task task, final PutDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        try {
+            Datasource jobParameter = Datasource.Builder.build(request);
+            IndexRequest indexRequest = new IndexRequest().index(DatasourceExtension.JOB_INDEX_NAME)
+                .id(jobParameter.getId())
+                .source(jobParameter.toXContent(JsonXContent.contentBuilder(), null))
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .opType(DocWriteRequest.OpType.CREATE);
+            client.index(indexRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(final IndexResponse indexResponse) {
+                    threadPool.generic().submit(() -> {
+                        try {
+                            createDatasource(jobParameter);
+                        } catch (Exception e) {
+                            log.error("Failed to create datasource for {}", jobParameter.getId(), e);
+                            jobParameter.getUpdateStats().setLastFailedAt(Instant.now());
+                            jobParameter.setState(DatasourceState.FAILED);
+                            try {
+                                DatasourceHelper.updateDatasource(client, jobParameter, timeout);
+                            } catch (Exception ex) {
+                                log.error("Failed to mark datasource state as FAILED for {}", jobParameter.getId(), ex);
+                            }
+                        }
+                    });
+                    listener.onResponse(new AcknowledgedResponse(true));
+                }
+
+                @Override
+                public void onFailure(final Exception e) {
+                    if (e instanceof VersionConflictEngineException) {
+                        log.info("Datasource already exists {}", request.getDatasourceName(), e);
+                        listener.onFailure(new ResourceAlreadyExistsException("Datasource already exists"));
+                    } else {
+                        log.error("Failed to create a datasource {}", request.getDatasourceName(), e);
+                        listener.onFailure(new OpenSearchException("Failed to create a datasource"));
+                    }
+                }
+            });
+        } catch (Exception e) {
+            log.error("Error occurred while creating datasource {}", request.getDatasourceName(), e);
```

Comment:
if we are passing exception to failure handler why we are logging it here? I believe its the job of the handler to log the exception properly. With this exception will be logged twice.

Reply:
for 1 & 2 why we want to do them? Like what is the benefit?

Reply:
Discussed offline. Will remove the logging.